### ISSUE TITLE
Database exceptions can't be handled with @UnitOfWork

### ISSUE THREAD

**Issue by ogregoire** [dropwizard/dropwizard#1621]
I'm using Dropwizard 1.0.0-rc3.

A constraint violation occurs on the database level when `@UnitOfWork` tries to commit (I guess that's the moment, actually). Yet no exception is thrown when I have the hand on the code.

I would expect to be able to handle this error gracefully. [The documentation has nothing on the subject](http://www.dropwizard.io/0.9.2/docs/manual/hibernate.html).

You can reproduce with my (under development) application. Just **send twice** the following request to localhost:8080.

```
Method: POST
URL: http://localhost:8080/books
MediaType: application/json
Json: {"isbn":"978-0618640157","title":"The Lord of the rings"}
```

Relevant files:
- [`migrations.xml`](https://github.com/ogregoire/alexandria/blob/3a219efd4fe9b0d5dc887f26054eb7ee17e88721/src/main/resources/migrations.xml): puts some unique constraint on the table
- [`BookResource.java`](https://github.com/ogregoire/alexandria/blob/3a219efd4fe9b0d5dc887f26054eb7ee17e88721/src/main/java/be/ogregoire/alexandria/core/book/BookResource.java): the resource where no exception is thrown.
- [`BookDAO.java`](https://github.com/ogregoire/alexandria/blob/3a219efd4fe9b0d5dc887f26054eb7ee17e88721/src/main/java/be/ogregoire/alexandria/core/book/BookDAO.java): fairly simple AbstractDAO implementation

If I'm doing something wrong, please advise and/or redirect to the appropriate documentation page.

**Comment by arteam** [comment]
I believe you hit the same issue as #1604, which was introduced in Dropwizard 1.0.0-rc3. We have a proposed fix in #1620, but it still needs to reviewed. If the fix is correct, the issue will be fixed in Dropwizard 1.0.0.

For the being, you can try downgrade to Dropwizard 1.0.0-rc2.

**Comment by ogregoire** [comment]
Well, that was fast.

Indeed, downgrading to 1.0.0-rc2 provided me with a status 500 Internal Server Error, which I expect more.

**But** I still can't handle the exception instead of throwing that 500. Is there anything I should do to handle those?

**Comment by arteam** [comment]
You need to declare and register an exception mapper which will translate a Hibernate exception to an appropriate HTTP response. An example of such a mapper: [DataExceptionMapper](https://github.com/dropwizard/dropwizard/blob/master/dropwizard-hibernate/src/test/java/io/dropwizard/hibernate/DataExceptionMapper.java). You can register it explicitly with the `environment.jersey.register` call in your application.

**Comment by ogregoire** [comment]
I don't really like what you describe though. I think I'll add a layer between the resource and the DAO, and proxy that layer with `@UnitOfWork` (since it's apparently possible). But thanks for leading me there, and also thanks for the very fast answers as well! :)

**Comment by arteam** [comment]
>  I think I'll add a layer between the resource and the DAO, and proxy that layer with @UnitOfWork (since it's apparently possible)

Fair enough.

>  But thanks for leading me there, and also thanks for the very fast answers as well! :)

You are welcome! Also thanks for taking your time to submit the issue with a detailed description. Hopefully, it will help other users who may encounter with it.

**Comment by hasanmumin** [comment]
I think this issue is still ongoing in version 1.0.2.  i using [DataExceptionMapper](https://github.com/dropwizard/dropwizard/blob/master/dropwizard-hibernate/src/test/java/io/dropwizard/hibernate/DataExceptionMapper.java) but noting to change. This before upgrade [register code](https://github.com/robeio/robe/blob/DW1.0-migration/robe-admin/src/main/java/io/robe/admin/RobeApplication.java#L135) and this [exception mapper](https://github.com/robeio/robe/blob/DW1.0-migration/robe-common/src/main/java/io/robe/common/exception/RobeExceptionMapper.java)

**Comment by nickbabcock** [comment]
@hasanmumin if you can create a small reproducible example using 1.0.2, that would be wonderful, and I can re-open this issue. If the issue isn't exactly the same then create a new issue so it doesn't get lost in a closed issue.

**Comment by hasanmumin** [comment]
@nickbabcock Hi. Than you for your response.
if you have little time please follow below step and you can test.
1- Please clone : https://github.com/robeio/robe/tree/DW1.0-migration
2- Change robe.yml hibernate section for your properties.
3- Than run RobeApplication with server robe-admin/robe.yml 
4- Add breakpoint to [DataExceptionMapper](https://github.com/robeio/robe/blob/DW1.0-migration/robe-common/src/main/java/io/robe/common/exception/DataExceptionMapper.java)
5- goto http://127.0.0.1:8080/robe/languages/example from you browser.

Console log:
> ERROR [2016-10-24 09:07:55,346] org.hibernate.internal.SessionImpl: HHH000346: Error during managed flush [Validation failed for classes [io.robe.admin.hibernate.entity.Language] during persist time for groups [javax.validation.groups.Default, ]
List of constraint violations:[
	ConstraintViolationImpl{interpolatedMessage='may not be null', propertyPath=code, rootBeanClass=class io.robe.admin.hibernate.entity.Language, messageTemplate='{javax.validation.constraints.NotNull.message}'}
]]

Browser:
`Bad Request.`

And it pass the breakpoint.
[Here](https://github.com/robeio/robe/blob/DW1.0-migration/robe-admin/src/main/java/io/robe/admin/RobeApplication.java#L141) jersey register DataExceptionMapper .
[Here](https://github.com/robeio/robe/blob/DW1.0-migration/robe-admin/src/main/java/io/robe/admin/resources/LanguageResource.java#L35) example hibarnate excepiton resource for you.

**Comment by nickbabcock** [comment]
Sorry to do this to you, but I think I'll want a self-contained, minimally reproducible project, so that it is easier to pinpoint the problem

Thanks 😃

**Comment by robertfirek** [comment]
@nickbabcock I think this problem still exists. Example project: https://github.com/robertfirek/dropwizard-exception-mapping

Looks like problem is somewhere in the `UnitOfWorkAspect`. When transaction start, this aspect throws exception which is not wrapped with `MappableException`. Not-mappable exception always generate 500.

**Comment by ogregoire** [comment]
As being the OP, this issue has NOT been solved. That is confirmed. However a workaround, which I described in my previous post, is to add a layer in between

**Comment by nickbabcock** [comment]
Re-opening the issue.

@ogregoire you had a proposed solution in #1625, which it looks like you weren't quite satisfied with; have you given it anymore thought?

**Comment by ogregoire** [comment]
@nickbabcock I added a new layer in-between to handle the exceptions myself.

So I faced other problems. I created #1625 to work on this, but in the end, it was a miserable failure because of the Guice integration. I then created #1661 which allowed me to integrate dropwizard-hibernate's  @UnitOfWork mechanism directly in Guice.

I still use that intermediary layer. Some might consider it a workaround, I don't as I get more control. Though the initial problem stays. Basic and "normal" usage of @UnitOfWork is broken.
